src/mscorlib/src/System/IO/StreamReader.cs

```diff
@@ -237,7 +244,23 @@ internal StreamReader(String path, Encoding encoding, bool detectEncodingFromByt
             byteLen = 0;
             bytePos = 0;
             _detectEncoding = detectEncodingFromByteOrderMarks;
-            _preamble = encoding.GetPreamble();
+
+            // Encoding.GetPreamble() always allocates and returns a new byte[] array for
+            // encodings that have a preamble.
+            // We can avoid repeated allocations for the default and commonly used Encoding.UTF8
+            // encoding by using our own private cached instance of the UTF8 preamble.
+            // We specifically look for Encoding.UTF8 because we know it has a preamble,
+            // whereas other instances of UTF8Encoding may not have a preamble enabled, and
+            // there's no public way to tell if the preamble is enabled for an instance other
+            // than calling GetPreamble(), which we're trying to avoid.
+            // This means that other instances of UTF8Encoding are excluded from this optimization.
+            // We check if the encoding is UTF8Encoding before the reference equality comparison
+            // with Encoding.UTF8, to avoid the potential first-time Encoding.UTF8 allocation
+            // when the encoding isn't even UTF8Encoding.
+            _preamble = (encoding is UTF8Encoding && object.ReferenceEquals(encoding, Encoding.UTF8)) ?
```

Comment:
I tend to agree with you, but wasn't sure if anyone would have any concern that this would cause `Encoding.UTF8` to be allocated (if it wasn't already) when some other encoding is used.